op-challenger/fault/responder.go

```diff
@@ -0,0 +1,163 @@
+package fault
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum-optimism/optimism/op-bindings/bindings"
+	"github.com/ethereum-optimism/optimism/op-service/txmgr"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// faultResponder implements the [Responder] interface to send onchain transactions.
+type faultResponder struct {
+	log log.Logger
+
+	shutdownCtx       context.Context
+	cancelShutdownCtx context.CancelFunc
+
+	// Transaction manager and configuration.
+	txMgr                  txmgr.TxManager
+	MaxPendingTransactions uint64
+
+	// Fault Dispute Game contract.
+	fdgAddr     *common.Address
+	fdgContract *bindings.FaultDisputeGameTransactor
+}
+
+// NewFaultResponder returns a new [faultResponder].
+func NewFaultResponder(logger log.Logger, fdgAddress *common.Address, fdgContract *bindings.FaultDisputeGameTransactor) faultResponder {
+	shutdownCtx, cancelShutdownCtx := context.WithCancel(context.Background())
+	return faultResponder{
+		shutdownCtx:       shutdownCtx,
+		cancelShutdownCtx: cancelShutdownCtx,
+		log:               logger,
+		fdgAddr:           fdgAddress,
+		fdgContract:       fdgContract,
+	}
+}
+
+// Quit shuts down the [faultResponder].
+func (r *faultResponder) Quit() {
+	r.log.Info("Shutting down fault responder")
+	r.cancelShutdownCtx()
+}
+
+// buildTxData builds the transaction data for the [faultResponder].
+func (r *faultResponder) buildTxData(ctx context.Context, response Claim) (TxData, error) {
+	// Get the parent index as a big integer.
+	parentIndex := response.Parent.ToGIndex()
+	bigParentIndex := big.NewInt(int64(parentIndex))
+
+	// Build the transaction data using the response claim.
+	if response.DefendsParent() {
+		txdata, err := r.fdgContract.Defend(&bind.TransactOpts{}, bigParentIndex, response.ValueBytes())
+		if err != nil {
+			return nil, err
+		}
+		return NewFaultResponderTxData(txdata), nil
+	} else {
+		txdata, err := r.fdgContract.Attack(&bind.TransactOpts{}, bigParentIndex, response.ValueBytes())
+		if err != nil {
+			return nil, err
+		}
+		return NewFaultResponderTxData(txdata), nil
+	}
+}
+
+// Respond takes a [Claim] and executes the response action.
+func (r *faultResponder) Respond(ctx context.Context, response Claim) error {
```

Comment:
Let's start with this function being single threaded & blocking. Doing the simplest thing means that we can wait for the transaction to be included on chain before performing the next move.

Reply:
not implemented yet

Reply:
I thought I did make it single threaded and blocking by waiting for the receipt?

Reply:
it's not returning the error from `Send`. Also there is no reason to have the tx queue at this point b/c it exists to send multiple transactions at once.